Sources/SwiftCompilerPlugin/Macros.swift

```diff
@@ -0,0 +1,254 @@
+  }
+
+  /// Expand `@freestainding(XXX)` macros.
+  func expandFreestandingMacro(
+    macro: PluginMessage.MacroReference,
+    discriminator: String,
+    expandingSyntax: PluginMessage.Syntax
+  ) throws {
+    let sourceManager = SourceManager()
+    let context = PluginMacroExpansionContext(
+      sourceManager: sourceManager,
+      discriminator: discriminator
+    )
+
+    let syntax = sourceManager.add(expandingSyntax)
+    guard let macroSyntax = syntax.asProtocol(FreestandingMacroExpansionSyntax.self) else {
+      let diag = PluginMessage.Diagnostic(
+        from: .init(node: syntax, message: MacroExpansionError.freestandingMacroSyntaxIsNotMacro),
+        in: sourceManager
+      )
+      try pluginHostConnection.sendMessage(
+        .expandFreestandingMacroResult(
+          expandedSource: "",
+          diagnostics: [diag]
+        )
+      )
+      return
+    }
+
+    let macroDef = self.resolveMacro(macro)
+    ///
+    if let exprMacroDef = macroDef as? ExpressionMacro.Type {
+      let rewritten = try exprMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    if let declMacroDef = macroDef as? DeclarationMacro.Type {
+      let rewritten = try declMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    let diag = PluginMessage.Diagnostic(
+      from: Diagnostic(node: syntax, message: MacroExpansionError.macroTypeNotFound),
+      in: sourceManager
+    )
+
+    try pluginHostConnection.sendMessage(
+      .expandFreestandingMacroResult(expandedSource: "", diagnostics: [diag])
+    )
+  }
+
+  /// Expand `@attached(XXX)` macros.
+  func expandAttachedMacro(
+    macro: PluginMessage.MacroReference,
+    macroRole: PluginMessage.MacroRole,
+    discriminator: String,
+    customAttributeSyntax: PluginMessage.Syntax,
+    declSyntax: PluginMessage.Syntax,
+    parentDeclSyntax: PluginMessage.Syntax?
+  ) throws {
+    let sourceManager = SourceManager()
+    let context = PluginMacroExpansionContext(
+      sourceManager: sourceManager,
+      discriminator: discriminator
+    )
+
+    guard let macroDefinition = resolveMacro(macro) else {
+      fatalError("macro type not found: \(macro.moduleName).\(macro.typeName)")
```

Comment:
Could we emit a diagnostic here similar to freestanding macros? Alternatively, I think this should have `FIXME` to highlight we need to fix this in the future.

Same for quite a few `fatalError`s below.